tests/instrumentation/pymemcache_tests.py

```diff
@@ -0,0 +1,150 @@
+
+pymemcache = pytest.importorskip("pymemcache")  # isort:skip
+
+
+pytestmark = [pytest.mark.pymemcache]
+
+
+if "MEMCACHED_HOST" not in os.environ:
+    pytestmark.append(pytest.mark.skip("Skipping pymemcache tests, no MEMCACHED_HOST environment variable set"))
+
+
+@pytest.mark.integrationtest
+def test_pymemcache_client(instrument, elasticapm_client):
+    elasticapm_client.begin_transaction("transaction.test")
+    host = os.environ.get("MEMCACHED_HOST", "localhost")
+    with capture_span("test_pymemcache", "test"):
+        conn = pymemcache.client.base.Client((host, 11211))
+        conn.set("mykey", "a")
+        assert b"a" == conn.get("mykey")
+        assert {"mykey": b"a"} == conn.get_many(["mykey", "myotherkey"])
+    elasticapm_client.end_transaction("BillingView")
+
+    transactions = elasticapm_client.events[TRANSACTION]
+    spans = elasticapm_client.spans_for_transaction(transactions[0])
+
+    expected_signatures = {"test_pymemcache", "Client.set", "Client.get", "Client.get_many"}
+
+    assert {t["name"] for t in spans} == expected_signatures
+
+    assert spans[0]["name"] == "Client.set"
+    assert spans[0]["type"] == "cache"
+    assert spans[0]["subtype"] == "memcached"
+    assert spans[0]["action"] == "query"
+    assert spans[0]["parent_id"] == spans[3]["id"]
+
+    assert spans[1]["name"] == "Client.get"
+    assert spans[1]["type"] == "cache"
+    assert spans[1]["subtype"] == "memcached"
+    assert spans[1]["action"] == "query"
+    assert spans[1]["parent_id"] == spans[3]["id"]
+    assert spans[1]["context"]["destination"] == {
+        "address": host,
+        "port": 11211,
+        "service": {"name": "memcached", "resource": "memcached", "type": "cache"},
+    }
+
+    assert spans[2]["name"] == "Client.get_many"
+    assert spans[2]["type"] == "cache"
+    assert spans[2]["subtype"] == "memcached"
+    assert spans[2]["action"] == "query"
+    assert spans[2]["parent_id"] == spans[3]["id"]
+
+    assert spans[3]["name"] == "test_pymemcache"
+    assert spans[3]["type"] == "test"
+
+    assert len(spans) == 4
+
+
+@pytest.mark.integrationtest
+def test_pymemcache_pooled_client(instrument, elasticapm_client):
+    elasticapm_client.begin_transaction("transaction.test")
+    host = os.environ.get("MEMCACHED_HOST", "localhost")
+    with capture_span("test_pymemcache", "test"):
+        conn = pymemcache.client.base.PooledClient((host, 11211))
+        conn.set("mykey", "a")
+        assert b"a" == conn.get("mykey")
+        assert {"mykey": b"a"} == conn.get_many(["mykey", "myotherkey"])
+    elasticapm_client.end_transaction("BillingView")
+
+    transactions = elasticapm_client.events[TRANSACTION]
+    spans = elasticapm_client.spans_for_transaction(transactions[0])
+
+    expected_signatures = {
+        "test_pymemcache",
+        "PooledClient.set",
+        "PooledClient.get",
+        "PooledClient.get_many",
+        "Client.set",
+        "Client.get",
+        "Client.get_many",
+    }
```

Comment:
Interestingly, no matter which class is used, all the "work" is done in a `Client` class at the bottom of the call chain. Despite this, I decided to instrument `PooledClient` and `HashClient` -- if they misbehave in the code surrounding the `Client` call, I think the user would want to know about it. That said, it does result in somewhat redundant spans.

Reply:
you could pass `leaf=True` to the `capture_span` calls, that would make it ignore child spans. Not sure if that's what you're alluding to, though

Reply:
I decided to use `leaf=True` for `PooledClient`, but not for `HashClient`. The reason for this is that `HashClient` deals with memcached clusters, and thus multiple `Client` objects per operation, while `PooledClient` only calls a single `Client` from the pool per operation, so the additional spans don't tell much.